src/main/java/dev/jbang/source/ProjectBuilder.java

```diff
@@ -258,12 +258,41 @@ private List<MavenRepo> allToMavenRepo(List<String> repos) {
 	}
 
 	public Project build(String resource) {
-		ResourceRef resourceRef = resolveChecked(getResourceResolver(), resource);
+		ResourceRef resourceRef = resolveChecked(getResourceResolver(), resource, true);
 		return build(resourceRef);
 	}
 
 	private ResourceRef resolveChecked(ResourceResolver resolver, String resource) {
-		Util.verboseMsg("Resolving resource ref: " + resource);
+		return resolveChecked(resolver, resource, false);
+	}
+
+	/**
+	 * Resolves the given resource, with property resolution from project context
+	 * and handles retry on caching.
+	 * 
+	 * Note: only enable property replacement if you know the resource is not
+	 * something that can be embedded/included directly in the project from users
+	 * data.
+	 * 
+	 * i.e. jbang run someref${os.detected.os}.jar is ok but //FILES with
+	 * ${user.home} is definitely not.
+	 * 
+	 * @param resolver
+	 * @param resource
+	 * @return
+	 */
+	private ResourceRef resolveChecked(ResourceResolver resolver, String orginal_resource,
+			boolean propertyReplacement) {
+		final String resource = propertyReplacement
+				? PropertiesValueResolver.replaceProperties(orginal_resource, getContextProperties())
+				: orginal_resource;
+
+		if (!resource.equals(orginal_resource)) {
+			Util.verboseMsg("Resolving resource ref: " + orginal_resource + " -> " + resource);
```

Comment:
Logging this is good, but I think the wording could be improved. In this case if there are no properties to be replaced you won't get a "resolving ..." message, which is wrong because we definitely are resolving. So I'd say that we'd either have to _always_ log something or change the wording to something like "Applying properties" instead of "Resolving".